core/src/main/java/org/kohsuke/stapler/export/ClassAttributeBehaviour.java

```diff
@@ -0,0 +1,103 @@
+package org.kohsuke.stapler.export;
+
+import org.jvnet.tiger_types.Types;
+
+import java.lang.reflect.Type;
+
+/**
+ * Setting that controls how the '_class' attribute will be produced in the output.
+ *
+ * <p>
+ * Three basic constants are defined:
+ *
+ * <dl>
+ *     <dt>{@link #NONE}
+ *     <dd>No type information will show up in the output whatsoever
+ *
+ *     <dt>{@link #ALWAYS}
+ *     <dd>Every object gets the type information explicitly written out
+ *
+ *     <dt>{@link #IF_NEEDED}
+ *     <dd>Type information will be produced only when it is necessary,
+ *         for example when the declared type and the actual type differ.
+ * </dl>
+ *
+ * <p>
+ * In addition, {@link #simple()} can be used to produce just the simple name of the class,
+ * whereas by default {@linkplain Class#getName() full class name} will be printed.
+ *
+ * @author Kohsuke Kawaguchi
+ * @see ExportConfig#withClassAttribute(ClassAttributeBehaviour)
+ */
+public abstract class ClassAttributeBehaviour {
+    private final String name;
+
+    // no subtyping outside this package
+    private ClassAttributeBehaviour(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString()+'['+name+']';
+    }
+
+    /**
+     * Determines if this type attribute should be written.
+     *
+     * @see DataWriter#type(Type, Class)
+     */
+    abstract Class map(Type expected, Class actual);
+
+    String print(Type expected, Class actual) {
```

Comment:
:ant: to be clear, please add `@CheckForNull` to each parameter and the return value, since IIUC all can be null.